File: ios/lib/features/chat/presentation/widgets/instagram_style_app_bar.dart
import 'package:flutter/material.dart';
import 'chat_theme.dart';

class InstagramStyleAppBar extends StatefulWidget implements PreferredSizeWidget {
  final String signalTitle;
  final List<String> participantAvatars;
  final SignalStatus status;
  final int onlineCount;
  final VoidCallback? onInfoPressed;
  final VoidCallback? onMenuPressed;

  const InstagramStyleAppBar({
    Key? key,
    required this.signalTitle,
    required this.participantAvatars,
    required this.status,
    this.onlineCount = 0,
    this.onInfoPressed,
    this.onMenuPressed,
  }) : super(key: key);

  @override
  State<InstagramStyleAppBar> createState() => _InstagramStyleAppBarState();

  @override
  Size get preferredSize => const Size.fromHeight(kToolbarHeight + 8);
}

class _InstagramStyleAppBarState extends State<InstagramStyleAppBar>
    with TickerProviderStateMixin {
  late AnimationController _pulseController;
  late Animation<double> _pulseAnimation;

  @override
  void initState() {
    super.initState();
    _setupAnimations();
  }

  void _setupAnimations() {
    _pulseController = AnimationController(
      duration: const Duration(milliseconds: 1500),
      vsync: this,
    );
    
    _pulseAnimation = Tween<double>(
      begin: 0.8,
      end: 1.2,
    ).animate(CurvedAnimation(
      parent: _pulseController,
      curve: Curves.easeInOut,
    ));

    _pulseController.repeat(reverse: true);
  }

  @override
  void dispose() {
    _pulseController.dispose();
    super.dispose();
  }

  Color _getStatusColor() {
    switch (widget.status) {
      case SignalStatus.active:
        return ChatColors.online;
      case SignalStatus.meeting:
        return ChatColors.accent;
      case SignalStatus.completed:
        return ChatColors.primary;
      case SignalStatus.expired:
      case SignalStatus.closed:
        return ChatColors.offline;
    }
  }

  String _getStatusText() {
    switch (widget.status) {
      case SignalStatus.active:
        return '활성 중';
      case SignalStatus.meeting:
        return '모임 진행 중';
      case SignalStatus.completed:
        return '모임 완료';
      case SignalStatus.expired:
        return '만료됨';
      case SignalStatus.closed:
        return '종료됨';
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      decoration: const BoxDecoration(
        color: ChatColors.surface,
        boxShadow: ChatShadows.card,
      ),
      child: SafeArea(
        child: Padding(
          padding: const EdgeInsets.symmetric(
            horizontal: ChatDimensions.paddingMD,
            vertical: ChatDimensions.paddingSM,
          ),
          child: Row(
            children: [
              // 뒤로가기 버튼
              IconButton(
                onPressed: () => Navigator.pop(context),
                icon: const Icon(
                  Icons.arrow_back_ios,
                  color: ChatColors.textPrimary,
                  size: 20,
                ),
                padding: EdgeInsets.zero,
                constraints: const BoxConstraints(),
              ),
              
              const SizedBox(width: ChatDimensions.paddingSM),
              
              // 참여자 프로필 원형 배열
              _buildParticipantProfiles(),
              
              const SizedBox(width: ChatDimensions.paddingMD),
              
              // 제목 및 상태 정보
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    Text(
                      widget.signalTitle,
                      style: ChatTextStyles.appBarTitle,
                      maxLines: 1,
                      overflow: TextOverflow.ellipsis,
                    ),
                    const SizedBox(height: 2),
                    Row(
                      children: [
                        // 상태 인디케이터
                        AnimatedBuilder(
                          animation: _pulseAnimation,
                          builder: (context, child) {
                            return Transform.scale(
                              scale: widget.status == SignalStatus.meeting 
                                ? _pulseAnimation.value 
                                : 1.0,
                              child: Container(
                                width: 8,
                                height: 8,
                                decoration: BoxDecoration(
                                  color: _getStatusColor(),
                                  shape: BoxShape.circle,
                                ),
                              ),
                            );
                          },
                        ),
                        const SizedBox(width: ChatDimensions.paddingSM),
                        Text(
                          _getStatusText(),
                          style: ChatTextStyles.appBarSubtitle.copyWith(
                            color: _getStatusColor(),
                          ),
                        ),
                        if (widget.onlineCount > 0) ...[
                          const SizedBox(width: ChatDimensions.paddingSM),
                          Text(
                            '• ${widget.onlineCount}명 온라인',
                            style: ChatTextStyles.appBarSubtitle,
                          ),
                        ],
                      ],
                    ),
                  ],
                ),
              ),
              
              // 액션 버튼들
              Row(
                mainAxisSize: MainAxisSize.min,
                children: [
                  if (widget.onInfoPressed != null)
                    IconButton(
                      onPressed: widget.onInfoPressed,
                      icon: const Icon(
                        Icons.info_outline,
                        color: ChatColors.textSecondary,
                        size: 22,
                      ),
                      padding: const EdgeInsets.all(ChatDimensions.paddingSM),
                      constraints: const BoxConstraints(),
                    ),
                  
                  if (widget.onMenuPressed != null)
                    IconButton(
                      onPressed: widget.onMenuPressed,
                      icon: const Icon(
                        Icons.more_vert,
                        color: ChatColors.textSecondary,
                        size: 22,
                      ),
                      padding: const EdgeInsets.all(ChatDimensions.paddingSM),
                      constraints: const BoxConstraints(),
                    ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget _buildParticipantProfiles() {
    const int maxVisible = 4;
    final visibleAvatars = widget.participantAvatars.take(maxVisible).toList();
    final remainingCount = widget.participantAvatars.length - maxVisible;

    return SizedBox(
      height: ChatDimensions.avatarMD,
      child: Stack(
        children: [
          // 프로필 아바타들
          ...visibleAvatars.asMap().entries.map((entry) {
            final index = entry.key;
            final avatarUrl = entry.value;
            
            return Positioned(
              left: index * 20.0, // 20px씩 겹치게 배치
              child: _buildProfileAvatar(avatarUrl, index),
            );
          }),
          
          // +N 표시 (남은 참여자 수)
          if (remainingCount > 0)
            Positioned(
              left: visibleAvatars.length * 20.0,
              child: _buildRemainingCount(remainingCount),
            ),
        ],
      ),
    );
  }

  Widget _buildProfileAvatar(String avatarUrl, int index) {
    return Container(
      width: ChatDimensions.avatarMD,
      height: ChatDimensions.avatarMD,
      decoration: BoxDecoration(
        shape: BoxShape.circle,
        border: Border.all(
          color: ChatColors.surface,
          width: 2,
        ),
        boxShadow: ChatShadows.card,
      ),
      child: CircleAvatar(
        radius: ChatDimensions.avatarMD / 2 - 2,
        backgroundColor: ChatColors.primary.withOpacity(0.1),
        backgroundImage: avatarUrl.isNotEmpty 
          ? NetworkImage(avatarUrl)
          : null,
        child: avatarUrl.isEmpty 
          ? Icon(
              Icons.person,
              color: ChatColors.primary,
              size: ChatDimensions.avatarMD / 2,
            )
          : null,
      ),
    );
  }

  Widget _buildRemainingCount(int count) {
    return Container(
      width: ChatDimensions.avatarMD,
      height: ChatDimensions.avatarMD,
      decoration: BoxDecoration(
        shape: BoxShape.circle,
        color: ChatColors.textSecondary,
        border: Border.all(
          color: ChatColors.surface,
          width: 2,
        ),
        boxShadow: ChatShadows.card,
      ),
      child: Center(
        child: Text(
          '+$count',
          style: const TextStyle(
            color: Colors.white,
            fontSize: 11,
            fontWeight: FontWeight.w600,
          ),
        ),
      ),
    );
  }
}

enum SignalStatus {
  active,
  meeting,
  completed,
  expired,
  closed,
}

File: android/app/src/main/java/com/signal/app/ui/screens/chat/components/InstagramStyleAppBar.kt
package com.signal.app.ui.screens.chat.components

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.zIndex
import coil.compose.AsyncImage
import com.signal.app.data.model.chat.SignalStatus
import com.signal.app.ui.theme.ChatColors
import com.signal.app.ui.theme.ChatDimensions
import com.signal.app.ui.theme.ChatShadows
import com.signal.app.ui.theme.ChatTextStyles

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun InstagramStyleAppBar(
    signalTitle: String,
    participantAvatars: List<String>,
    status: SignalStatus,
    onlineCount: Int = 0,
    onBackPressed: () -> Unit = {},
    onInfoPressed: (() -> Unit)? = null,
    onMenuPressed: (() -> Unit)? = null,
    modifier: Modifier = Modifier
) {
    val infiniteTransition = rememberInfiniteTransition(label = "pulse")
    val pulseScale by infiniteTransition.animateFloat(
        initialValue = 0.8f,
        targetValue = 1.2f,
        animationSpec = infiniteRepeatable(
            animation = tween(1500, easing = LinearEasing),
            repeatMode = RepeatMode.Reverse
        ),
        label = "pulse_scale"
    )

    TopAppBar(
        title = {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.fillMaxWidth()
            ) {
                // Participant profiles
                ParticipantProfiles(
                    avatars = participantAvatars,
                    modifier = Modifier.padding(end = ChatDimensions.PaddingMD)
                )
                
                // Title and status
                Column(
                    modifier = Modifier.weight(1f)
                ) {
                    Text(
                        text = signalTitle,
                        style = ChatTextStyles.AppBarTitle,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis
                    )
                    
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        modifier = Modifier.padding(top = 2.dp)
                    ) {
                        // Status indicator
                        Box(
                            modifier = Modifier
                                .size(8.dp)
                                .scale(
                                    if (status == SignalStatus.MEETING) pulseScale else 1f
                                )
                                .background(
                                    color = getStatusColor(status),
                                    shape = CircleShape
                                )
                        )
                        
                        Spacer(modifier = Modifier.width(ChatDimensions.PaddingSM))
                        
                        Text(
                            text = getStatusText(status),
                            style = ChatTextStyles.AppBarSubtitle.copy(
                                color = getStatusColor(status)
                            )
                        )
                        
                        if (onlineCount > 0) {
                            Text(
                                text = " • ${onlineCount}명 온라인",
                                style = ChatTextStyles.AppBarSubtitle
                            )
                        }
                    }
                }
            }
        },
        navigationIcon = {
            IconButton(onClick = onBackPressed) {
                Icon(
                    imageVector = Icons.Default.ArrowBack,
                    contentDescription = "뒤로가기",
                    tint = ChatColors.TextPrimary
                )
            }
        },
        actions = {
            onInfoPressed?.let {
                IconButton(onClick = it) {
                    Icon(
                        imageVector = Icons.Default.Info,
                        contentDescription = "정보",
                        tint = ChatColors.TextSecondary
                    )
                }
            }
            
            onMenuPressed?.let {
                IconButton(onClick = it) {
                    Icon(
                        imageVector = Icons.Default.MoreVert,
                        contentDescription = "메뉴",
                        tint = ChatColors.TextSecondary
                    )
                }
            }
        },
        colors = TopAppBarDefaults.topAppBarColors(
            containerColor = ChatColors.Surface
        ),
        modifier = modifier.shadow(
            elevation = ChatShadows.Card,
            spotColor = ChatShadows.SoftColor
        )
    )
}

@Composable
private fun ParticipantProfiles(
    avatars: List<String>,
    modifier: Modifier = Modifier
) {
    val maxVisible = 4
    val visibleAvatars = avatars.take(maxVisible)
    val remainingCount = avatars.size - maxVisible

    Box(
        modifier = modifier.height(ChatDimensions.AvatarMD)
    ) {
        visibleAvatars.forEachIndexed { index, avatarUrl ->
            ProfileAvatar(
                avatarUrl = avatarUrl,
                modifier = Modifier
                    .offset(x = (index * 20).dp)
                    .zIndex((maxVisible - index).toFloat())
            )
        }
        
        if (remainingCount > 0) {
            RemainingCountAvatar(
                count = remainingCount,
                modifier = Modifier
                    .offset(x = (visibleAvatars.size * 20).dp)
                    .zIndex(0f)
            )
        }
    }
}

@Composable
private fun ProfileAvatar(
    avatarUrl: String,
    modifier: Modifier = Modifier
) {
    Box(
        modifier = modifier
            .size(ChatDimensions.AvatarMD)
            .border(
                width = 2.dp,
                color = ChatColors.Surface,
                shape = CircleShape
            )
            .shadow(
                elevation = ChatShadows.Card,
                shape = CircleShape,
                spotColor = ChatShadows.SoftColor
            )
    ) {
        if (avatarUrl.isNotEmpty()) {
            AsyncImage(
                model = avatarUrl,
                contentDescription = "프로필 이미지",
                contentScale = ContentScale.Crop,
                modifier = Modifier
                    .fillMaxSize()
                    .clip(CircleShape)
            )
        } else {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .background(
                        color = ChatColors.Primary.copy(alpha = 0.1f),
                        shape = CircleShape
                    ),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    painter = painterResource(android.R.drawable.ic_menu_my_calendar),
                    contentDescription = null,
                    tint = ChatColors.Primary,
                    modifier = Modifier.size((ChatDimensions.AvatarMD / 2))
                )
            }
        }
    }
}

@Composable
private fun RemainingCountAvatar(
    count: Int,
    modifier: Modifier = Modifier
) {
    Box(
        modifier = modifier
            .size(ChatDimensions.AvatarMD)
            .background(
                color = ChatColors.TextSecondary,
                shape = CircleShape
            )
            .border(
                width = 2.dp,
                color = ChatColors.Surface,
                shape = CircleShape
            )
            .shadow(
                elevation = ChatShadows.Card,
                shape = CircleShape,
                spotColor = ChatShadows.SoftColor
            ),
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = "+$count",
            color = ChatColors.TextOnPrimary,
            fontSize = 11.sp,
            style = ChatTextStyles.QuickReplyText
        )
    }
}

private fun getStatusColor(status: SignalStatus): Color {
    return when (status) {
        SignalStatus.ACTIVE -> ChatColors.Online
        SignalStatus.MEETING -> ChatColors.Accent
        SignalStatus.COMPLETED -> ChatColors.Primary
        SignalStatus.EXPIRED, SignalStatus.CLOSED -> ChatColors.Offline
    }
}

private fun getStatusText(status: SignalStatus): String {
    return when (status) {
        SignalStatus.ACTIVE -> "활성 중"
        SignalStatus.MEETING -> "모임 진행 중"
        SignalStatus.COMPLETED -> "모임 완료"
        SignalStatus.EXPIRED -> "만료됨"
        SignalStatus.CLOSED -> "종료됨"
    }
}